configurate-core/src/main/java/ninja/leaping/configurate/objectmapping/FieldAdapter.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Configurate
+ * Copyright (C) zml and Configurate contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ninja.leaping.configurate.objectmapping;
+
+import ninja.leaping.configurate.ConfigurationNode;
+
+public interface FieldAdapter<V> {
```

Comment:
Why not just use the `TypeSerializer` interface instead? I realise that it also includes the `TypeToken<?>` parameter, but it then allows the same logic to be used in a `TypeSerializerCollection` (perhaps for child objects) as well as a field adapter.

Reply:
From the naming convention when i see the class  "TypeSerializer" for the first time i will automatically make an assumption that the typeserializer shall be used only for one type only. If theres a "Double" field the TypeSerializer<Double> will always be used etc. 

FieldAdapter as now exists may be used on different fields with different types, if the user wishes to. Therefore i belive that adding parameter TypeToken would only add to the confusion.

Reply:
> From the naming convention when i see the class "TypeSerializer" for the first time i will automatically make an assumption that the typeserializer shall be used only for one type only. If theres a "Double" field the TypeSerializer<Double> will always be used etc.

It's already not the case now. Some `TypeSerializer`s operate on fields that might not necessarily be their concrete type. [We do that with some types in Sponge, one example is `DataSerializer`s.](https://github.com/SpongePowered/SpongeCommon/blob/stable-7/src/main/java/org/spongepowered/common/data/SpongeDataManager.java#L81)

[A good example in stock Configurate is that a `TypeSerializer<Number>` is used for the `Double` field](https://github.com/SpongePowered/configurate/blob/1cbb1bb610cd1ba09229e3f7bc38a9bdbede6d66/configurate-core/src/main/java/ninja/leaping/configurate/objectmapping/serialize/TypeSerializers.java#L73).

If you think there is a documentation issue there, then feel free to flag it up.

> FieldAdapter as now exists may be used on different fields with different types, if the user wishes to.

That doesn't change here if it remains as a `TypeSerializer`, as discussed above.

> i will automatically make an assumption that the typeserializer shall be used only for one type only. 

I could make the same assumption about your proposal, as it has a type parameter that I might assume should match up with the field type exactly.

> Therefore i belive that adding parameter TypeToken would only add to the confusion.

The `TypeToken` simply provides the type of target field. It can be ignored, to checked for compatibility with the field allowing for better errors to be thrown if the adapter is used on an incompatible field. How is that going to confuse people? For some people writing custom serialisers, this is useful. It may well add flexibility to how people write their serialisers intended for the Adapter annotation.

---

My point is that you're still serialising/deserialising, all you are doing is _explicity_ selecting a serialisation method instead of letting Configurate choose for you. How is that different from what a `TypeSerializer` does? It's not. You are still going to ultimately be translating a specific type using the adapter. I don't think we should create a second interface that effectively does the same thing.

Ultimately, I don't think we need this `FieldAdapter`. Documentation on the annotation is needed to explain its use, possibly with some clarification on the use of the `TypeSerializer`.

Reply:
Adapters now use TypeSer. as well, see a1798168eb0801e45cf8ca47bec282a51a9e490c